File: src/simulation/space/resource/ResourceIdeal.java
package simulation.space.resource;

import simulation.culture.aspect.Aspect;
import simulation.space.tile.Tile;

import java.util.List;

/**
 * Special resource instance, which is never placed on the map and prints
 * a warning on any attempt of changing it, but which can be used as
 * an example instance of a resource inside classes working with resources.
 */
public class ResourceIdeal extends Resource {
    public ResourceIdeal(ResourceCore resourceCore) {
        super(resourceCore, 1);
    }

    @Override
    public Resource getPart(int part) {
        System.err.println("Ideal is changing");
        return super.getPart(part);
    }

    @Override
    public Resource merge(Resource resource) {
        System.err.println("Ideal is changing");
        return super.merge(resource);
    }

    @Override
    public ResourceUpdateResult update(Tile tile) {
        System.err.println("Ideal is changing");
        return super.update(tile);
    }

    @Override
    public void addAmount(int amount) {
        System.err.println("Ideal is changing");
        super.addAmount(amount);
    }

    @Override
    public List<Resource> applyAndConsumeAspect(Aspect aspect, int part, boolean isClean) {
        System.err.println("Ideal is changing");
        return super.applyAndConsumeAspect(aspect, part, isClean);
    }
}

File: EcosystemSimulation/src/io/tashtabash/simulation/space/resource/ResourceIdeal.kt
package io.tashtabash.simulation.space.resource

import io.tashtabash.simulation.space.tile.Tile
import javax.naming.OperationNotSupportedException

/**
 * Special resource instance, which must is never be placed on the map and
 * will print a warning on any attempt of changing it. Can be used as
 * an example instance of a resource inside classes working with resources.
 */
class ResourceIdeal(
        genome: Genome,
        amount: Int = 1,
        val copyFun: ((Int, Int) -> Resource)? = null
) : Resource(ResourceCore(genome, ArrayList(), freeMarker), amount) {
    override fun copy(amount: Int, deathTurn: Int): Resource {
        return copyFun?.invoke(amount, deathTurn)
                ?: super.copy(amount, deathTurn)
    }

    override fun getPart(part: Int, taker: Taker): Resource {
        throw OperationNotSupportedException("ResourceIdeal doesn't support getPart(..)")
    }

    override fun getCleanPart(part: Int, taker: Taker): Resource {
        throw OperationNotSupportedException("ResourceIdeal doesn't support applyActionAndConsume(..)")
    }

    override fun merge(resource: Resource): Resource {
        throw OperationNotSupportedException("ResourceIdeal doesn't support merge(..)")
    }

    override fun update(tile: Tile): ResourceUpdateResult {
        throw OperationNotSupportedException("ResourceIdeal doesn't support update(..)")
    }

    override fun addAmount(amount: Int) {
        throw OperationNotSupportedException("ResourceIdeal doesn't support addAmount(..)")
    }
}